pennylane/interfaces/tf.py

```diff
@@ -99,30 +132,17 @@ def grad(grad_output, **tfkwargs):
             """Returns the vector-Jacobian product"""
             # evaluate the Jacobian matrix of the QNode
             variables = tfkwargs.get("variables", None)
-
             jacobian = qnode.jacobian(args, kwargs)
-            grad_output_np = grad_output.numpy()
-
-            # perform the vector-Jacobian product
-            if not grad_output_np.shape:
-                temp = grad_output_np * jacobian
-            else:
-                temp = grad_output_np.T @ jacobian
-
-            # restore the nested structure of the input args
-            grad_input = unflatten(temp.flat, args)
-
-            if isinstance(grad_input, list):
-                grad_input = [tf.convert_to_tensor(i, dtype=dtype) for i in grad_input]
-            elif isinstance(grad_input, tuple):
-                grad_input = tuple(tf.convert_to_tensor(i, dtype=dtype) for i in grad_input)
-            else:
-                grad_input = tf.convert_to_tensor(grad_input, dtype=dtype)
+            jacobian = tf.constant(jacobian, dtype=dtype)
+            grad_output_row = tf.transpose(tf.reshape(grad_output, [-1, 1]))
+            grad_input = tf.matmul(grad_output_row, jacobian)
+            grad_input = tf.reshape(grad_input, [-1])
+            grad_input_unflattened = unflatten_tf(grad_input, input_)[0]
```

Comment:
I like that you got rid of the 'wrapper' `unflatten()` function! Always annoyed me that we had two. I wonder if it makes sense to keep the assertion in, however. Something like:
```suggestion
            grad_input_unflattened, tail = unflatten_tf(grad_input, input_)
            assert tail, "Flattened iterable has more elements than the model."
```

The reason to go for an assertion over an exception here is that:

1. This is an exceptional program state that _should not occur_, there is no logic available to recover from this.

2. It is not not a program state that can be conceivably caused by user input.

3. Assertions can be disabled during execution by the user if required as an optimization.

Reply:
Nice! I had trouble to get this to work when TF is in graph mode (in `test_qnode_jacobian_agrees`). The following would work, but I think it might now be overkill:
```python
@tf.function
def assert_no_tail(tail):
    size_zero = tf.equal(tf.size(tail), 0)
    tf.Assert(size_zero, [tf.constant("Flattened iterable has more elements than the "
                                      "model.")])

assert_no_tail(tail)
```
What do you think, worth including? Also, the printout from the TF assert looks bad.

Reply:
@trbromley what do you mean by 'in graph mode'? I don't think we can support graph mode with our tf interface, since it includes quantum hardware!

Reply:
:thinking: I'm not sure what it should be called, the time when the TF tensor is symbolic and doesn't have a `numpy()` method?

Reply:
Yep, I don't think that's supported! Was there a case where you had it working?

Reply:
Yep, the above function works also in symbolic mode, but you have to do everything in TF (`tf.equals`, `tf.Assert`) - I think we can just omit the assert for now due to these difficulties.